### PR TITLE
[LIE/HM] compute & output second varaibles

### PR DESCRIPTION
**Pull request by norihiro-w**
Changes in this PR is aimed at saving some internal variables (e.g. fracture force vector, fracture model state) in LIE/HM process into result VTU files. 

The last commit is formatting. I suggest reviewing commit wise

### CODE REVIEW

File: MaterialLib/FractureModels/FractureModelBase.h [hunk 1]
```diff
@@ -27,6 +27,38 @@ template <int DisplacementDim>
...
+        virtual ~MaterialStateVariables() = default;
+        virtual void pushBackState() = 0;
+
+        void reset()
+        {
+            _is_tensile_stress = false;
+            _shear_yield_function = 0.0;
+        }
+
+        void setShearYieldFucntion(double Fs) { _shear_yield_function = Fs; }
```
Comment (TomFischer) [inline MaterialLib/FractureModels/FractureModelBase.h]: Spelling: Function

File: MaterialLib/FractureModels/FractureModelBase.h [hunk 2]
```diff
@@ -27,6 +27,38 @@ template <int DisplacementDim>
...
+        double getShearYieldFucntion() const { return _shear_yield_function; }
```
Comment (TomFischer) [inline MaterialLib/FractureModels/FractureModelBase.h]: Spelling: Function

File: MaterialLib/FractureModels/FractureModelBase.h [hunk 3]
```diff
@@ -48,7 +80,8 @@ class FractureModelBase
             Eigen::Ref<Eigen::VectorXd const> w,
             Eigen::Ref<Eigen::VectorXd const> sigma_prev,
             Eigen::Ref<Eigen::VectorXd> sigma,
-            Eigen::Ref<Eigen::MatrixXd> C)  = 0;
+            Eigen::Ref<Eigen::MatrixXd> C,
+            MaterialStateVariables& material_state_variables)  = 0;
```
Comment (TomFischer) [inline MaterialLib/FractureModels/FractureModelBase.h]: The documentation for the new parameter is missing.

File: MaterialLib/FractureModels/FractureModelBase.h [hunk 4]
```diff
@@ -27,6 +27,38 @@ template <int DisplacementDim>
...
+        void reset()
+        {
+            _is_tensile_stress = false;
+            _shear_yield_function = 0.0;
+        }
+
+        void setShearYieldFucntion(double Fs) { _shear_yield_function = Fs; }
+        double getShearYieldFucntion() const { return _shear_yield_function; }
+
+        void isTensileStress(bool flag) { _is_tensile_stress = flag; }
```
Comment (TomFischer) [inline MaterialLib/FractureModels/FractureModelBase.h]: Using this method the flag is set. For this reason I would prefer a method name `setTensileStress`.

File: MaterialLib/FractureModels/LinearElasticIsotropic.cpp [hunk 5]
```diff
@@ -22,8 +22,12 @@ void LinearElasticIsotropic<DisplacementDim>::computeConstitutiveRelation(
         Eigen::Ref<Eigen::VectorXd const> w,
         Eigen::Ref<Eigen::VectorXd const> sigma_prev,
         Eigen::Ref<Eigen::VectorXd> sigma,
-        Eigen::Ref<Eigen::MatrixXd> C)
+        Eigen::Ref<Eigen::MatrixXd> C,
+        typename FractureModelBase<DisplacementDim>::MaterialStateVariables&
```
Comment (TomFischer) [inline MaterialLib/FractureModels/LinearElasticIsotropic.cpp]: Maybe `\copydoc` can help you?

File: ProcessLib/LIE/HydroMechanics/LocalAssembler/HydroMechanicsLocalAssemblerFracture-impl.h [hunk 6]
```diff
@@ -293,19 +293,99 @@ void
...
+        ele_sigma_eff += ip.sigma_eff;
+        ele_Fs = std::max(ele_Fs, ip.material_state_variables->getShearYieldFucntion());
     }
-    ele_b /= _ip_data.size();
-    ele_k /= _ip_data.size();
+    ele_b /= static_cast<double>(_ip_data.size());
+    ele_k /= static_cast<double>(_ip_data.size());
+    ele_w /= static_cast<double>(_ip_data.size());
+    ele_sigma_eff /= static_cast<double>(_ip_data.size());
     (*_process_data.mesh_prop_b)[this->_element.getID()] = ele_b;
```
Comment (TomFischer) [inline ProcessLib/LIE/HydroMechanics/LocalAssembler/HydroMechanicsLocalAssemblerFracture-impl.h]: Why not `auto const elem_id = this->_element.getID()` and use this `elem_id` in the following lines. This would be more readable IMHO.

File: ProcessLib/LIE/HydroMechanics/LocalAssembler/HydroMechanicsLocalAssemblerMatrix-impl.h [hunk 7]
```diff
@@ -270,8 +270,66 @@ void
...
+
+        auto& eps = ip_data.eps;
+        auto& sigma_eff = ip_data.sigma_eff;
+        auto& C = ip_data.C;
+        auto& material_state_variables = *ip_data.material_state_variables;
+        double const k_over_mu =
+            _process_data.intrinsic_permeability(t, x_position)[0] /
+            _process_data.fluid_viscosity(t, x_position)[0];
+        auto const rho_fr = _process_data.fluid_density(t, x_position)[0];
+        auto const& gravity_vec = _process_data.specific_body_force;
```
Comment (TomFischer) [inline ProcessLib/LIE/HydroMechanics/LocalAssembler/HydroMechanicsLocalAssemblerMatrix-impl.h]: I'm unsure. Maybe you can move it outside of the integration loop?

File: ProcessLib/LIE/SmallDeformation/LocalAssembler/IntegrationPointDataFracture.h [hunk 8]
```diff
@@ -26,11 +26,12 @@ struct IntegrationPointDataFracture final
...
     explicit IntegrationPointDataFracture(
         MaterialLib::Fracture::FractureModelBase<DisplacementDim>& fracture_material)
-        : _fracture_material(fracture_material)
+        : _fracture_material(fracture_material),
+          _material_state_variables(
+              _fracture_material.createMaterialStateVariables())
     {
     }
 
-#if defined(_MSC_VER) && _MSC_VER < 1900
```
Comment (endJunction) [inline ProcessLib/LIE/SmallDeformation/LocalAssembler/IntegrationPointDataFracture.h]: Just curious, why did you remove the ifdef?

File: ProcessLib/LIE/HydroMechanics/LocalAssembler/HydroMechanicsLocalAssemblerMatrixNearFracture-impl.h [hunk 9]
```diff
@@ -117,6 +117,44 @@ assembleWithJacobianConcrete(
...
+        fracture_props, _element.getCenterOfGravity().getCoords());
+
+    if (ele_levelset == 0)
+    {
+        // no DoF exists for displacement jumps. do the normal assebmly
+        Base::computeSecondaryVariableConcreteWithBlockVectors(t, p, u);
+        return;
+    }
+
+    // Displacement jumps should be taken into account
```
Comment (endJunction) [inline ProcessLib/LIE/HydroMechanics/LocalAssembler/HydroMechanicsLocalAssemblerMatrixNearFracture-impl.h]: Is it a todo, or are the displacement jumps taken into account below?

**Comment by endJunction** [comment]
:fast_forward:

File: ProcessLib/LIE/HydroMechanics/LocalAssembler/HydroMechanicsLocalAssemblerMatrixNearFracture-impl.h (hunk 9 above)
Comment (norihiro-w) [inline ProcessLib/LIE/HydroMechanics/LocalAssembler/HydroMechanicsLocalAssemblerMatrixNearFracture-impl.h]: no it's not todo. it says now the assembly should take into account the displacement jumps, i.e. need to compute true displacement and assemble r and J for the variables.
